### ISSUE TITLE
When TextField is input, the cursor is dancing

### ISSUE THREAD

**Issue by Geekoo-Joker** [flutter/flutter#32278]
TextField cursor position is abnormal，after I update.
When multi-line input, text and cursor will swing up and down.
And the line spacing of the text is too small, they are all superimposed together.

Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.5.4-hotfix.2, on Mac OS X 10.14.4 18E226, locale
    zh-Hans-CN)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 10.2.1)
[✓] Android Studio (version 3.4)
[✓] Connected device (1 available)

• No issues found!

**Comment by appxpgraph** [comment]
+1

**Comment by HansMuller** [comment]
Thanks for the report! If you can provide a small self-contained app that demonstrates the problem, we'll look into it.

**Comment by Geekoo-Joker** [comment]
TextField(
      //controller: text,
      //focusNode: focus,
      //cursorColor: Color_Main,
      style: TextStyle(
        //fontSize: 15 / MediaQuery.of(context).textScaleFactor,
        //color: Color_4,
        //height: 1.5,
      ),
      maxLines: null, // Here is important . Chinese. Android does not have this problem, only iOS has. 
      //autofocus: true,
      textInputAction: TextInputAction.send,
      decoration: InputDecoration(
        //hintText: hitText,
        contentPadding: EdgeInsets.all(5),
        hintStyle: TextStyle(
          //fontSize: 15 / MediaQuery.of(context).textScaleFactor,
          //color: Color_Hint,
          //height: 1.5,
        ),
        border: InputBorder.none,
      ),
      //onSubmitted: send,
    )


All textfield cursors are smaller than words.
Long press, there will be a full screen of "pasteButtonLabel was called on null". Because of this problem, my App was rejected by Apple.

**Comment by HansMuller** [comment]
CC @GaryQian

**Comment by HansMuller** [comment]
The test case appears to be equivalent to:
```
void main() {
  runApp(MaterialApp(
    home: Scaffold(
      body: Center(
        child: TextField(
          style: TextStyle(),
          maxLines: null, // Here is important . Chinese. Android does not have this problem, only iOS has.
          textInputAction: TextInputAction.send,
          decoration: InputDecoration(
            contentPadding: EdgeInsets.all(5),
            hintStyle: TextStyle(),
            border: InputBorder.none,
          ),
        ),
      ),
    ),
  ));
}
```
Is it really necessary to set the text styles to `TextStyle()`?

**Comment by HansMuller** [comment]
A message like `pasteButtonLabel was called on null` when the app is run on a device configured for Chinese implies that the app's cupertino or material localizations haven't been set up.  There's more about that here: https://flutter.dev/docs/development/accessibility-and-localization/internationalization

**Comment by GaryQian** [comment]
I have recently landed some caret size and positioning fixes.  See https://github.com/flutter/flutter/pull/31210 and https://github.com/flutter/flutter/pull/31687

We are aware that the caret's size can be strange and are actively working on improving it for edge cases.

Can you explain what you mean by:

>  text and cursor will swing up and down.

I'm not seeing behavior like that, at least it is not obvious.

**Comment by Geekoo-Joker** [comment]
@GaryQian 

I recorded some resources.

https://github.com/Geekoo-Joker/someResource/blob/master/1.jpg
This image records the code.

https://github.com/Geekoo-Joker/someResource/blob/master/1.mp4
This video corresponds to the question.

You can clearly see the cursor and text swing when you type in the first line.
The second line will not swing when you enter. At the same time you can see that the line spacing is too small. Especially when you input Chinese later. And the cursor is weird.


https://github.com/Geekoo-Joker/someResource/blob/master/2.jpg
This image is an improved code. I have to add a line height.

https://github.com/Geekoo-Joker/someResource/blob/master/2.mp4
This video is a usage experience.

It's a lot better on iOS now. But in Android, the upstream spacing is too large

**Comment by GaryQian** [comment]
Thanks for providing this info! However it looks like just your TextField widget is not enough to be able to repro this. You seem to be using a custom font that seems to be bundled into your TextField via the theme. Also, constraint details around your TextField may also impact how it behaves. Do you think you can provide a full flutter app that we can simply download/clone and run that will 100% repro your issue? Feel free to strip anything sensitive from it. Your code snippet by itself is not enough.

**Comment by GaryQian** [comment]
cc @justinmc This may have to do with your textfield height behavior change?

**Comment by GaryQian** [comment]
So after talking to some people about this, it seems that in your exact font configuration, the Strut is too short. Although this is still a bug on our end, you can try to work around it for now by providing a StrutStyle to the TextField with a slightly larger font size than the text itself.

**Comment by Geekoo-Joker** [comment]
@GaryQian 

https://github.com/Geekoo-Joker/reproduction
![image of 1](https://github.com/Geekoo-Joker/someResource/blob/master/1.png)

I think I have guessed the problem.
The text size of my iPhone is relatively large.
When I turn down the text of the iPhone, the textField basically works fine.

![image of IMG_3371](https://github.com/Geekoo-Joker/someResource/blob/master/IMG_3371.PNG)

**Comment by 100cm** [comment]
any progress on this?

**Comment by 100cm** [comment]
same problem here.

**Comment by GaryQian** [comment]
So it looks like the problem here is actually the strut is not scaling with the desired font size. If both the text's actual font size and the underlying strut were scaled at the same time, this would not happen.

**Comment by GaryQian** [comment]
The textScaleFactor is not being accounted for in strutStyle.

**Comment by GaryQian** [comment]
https://github.com/flutter/flutter/pull/33462 fixes this.
